File: src/main/java/core/scm/specialforms/Throw.java
package core.scm.specialforms;

import core.environment.Environment;
import core.evaluator.Evaluator;
import core.exceptions.IllegalSyntaxException;
import core.exceptions.ThrowableWrapper;
import core.exceptions.WrongTypeException;

import java.util.List;

public enum Throw implements ISpecialForm {
  THROW;

  private static final StackTraceElement[] EMPTY = new StackTraceElement[0];

  @Override
  public Object eval(List<Object> expression, Environment env, Evaluator evaluator) {
    if (expression.size() < 2) {
      throw IllegalSyntaxException.of(toString(), expression);
    }
    Object obj = evaluator.eval(expression.get(1), env);
    if (!(obj instanceof Throwable)) {
      throw new WrongTypeException(toString(), "Throwable", obj);
    }
    Throwable throwable = (Throwable) obj;
    /* Clear stacktrace */
    throwable.setStackTrace(EMPTY);
    throw new ThrowableWrapper(throwable);
  }

  @Override
  public String toString() {
    return "throw";
  }
}

File: src/main/java/core/scm/specialforms/Throw.kt
package core.scm.specialforms

import core.environment.Environment
import core.evaluator.Evaluator
import core.exceptions.IllegalSyntaxException
import core.exceptions.ThrowableWrapper
import core.exceptions.WrongTypeException

enum class Throw : ISpecialForm {
    THROW;

    override fun eval(expression: List<Any?>, env: Environment, evaluator: Evaluator): Any? {
        if (expression.size < 2) {
            throw IllegalSyntaxException.of(toString(), expression)
        }
        val obj = evaluator.eval(expression[1], env)
        if (obj !is Throwable) {
            throw WrongTypeException(toString(), "Throwable", obj)
        }
        throw ThrowableWrapper((obj as Throwable?)!!)
    }

    override fun toString(): String {
        return "throw"
    }
}